{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d778d6fd-e0f7-4965-97d0-d7349a4cd518",
   "metadata": {},
   "source": [
    "## Part 1. Running PRS-CSx for variants' weights"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ed6441e-ce96-47fb-9f95-f83b757b538a",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "1. reference panel for EAS population: ldblk_1kg_eas.tar.gz (from dropbox)\n",
    "2. reference panel for EUR population: ldblk_1kg_eur.tar.gz (from dropbox)\n",
    "3. SNP information: snpinfo_mult_1kg_hm3 (from dropbox)\n",
    "4. EAS population summary statistics for schizophrenia: EAS_sumstats.txt (from test data in PRS-CSx Github folder)\n",
    "5. EUR population summary statistics for schizophrenia: EUR_sumstats.txt (from test data in PRS-CSx Github folder)\n",
    "\n",
    "### Outputs\n",
    "1. EAS PRS (contains population-specific posterior SNP effect size estimates for each individual): test_EAS_pst_eff_a1_b0.5_phi1e-02_chr22.txt\n",
    "2. EUR PRS (contains population-specific posterior SNP effect size estimates for each individual): test_EUR_pst_eff_a1_b0.5_phi1e-02_chr22.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78ceac7c-3547-49ad-b76f-c93615af18a9",
   "metadata": {},
   "source": [
    "### Check if bash and curl are available:\n",
    "If you run into an error, please install them and then run the following"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d99980dd-d372-46a2-94f6-dd5263ca82f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GNU bash, version 3.2.57(1)-release (arm64-apple-darwin23)\n",
      "Copyright (C) 2007 Free Software Foundation, Inc.\n",
      "/bin/bash\n",
      "curl 8.9.1 (arm64-apple-darwin20.0.0) libcurl/8.9.1 OpenSSL/3.0.15 zlib/1.2.13 libssh2/1.11.0 nghttp2/1.57.0\n",
      "Release-Date: 2024-07-31\n",
      "Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp\n",
      "Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz NTLM SPNEGO SSL threadsafe TLS-SRP UnixSockets\n"
     ]
    }
   ],
   "source": [
    "!bash --version\n",
    "!which bash\n",
    "!curl --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58be98c1",
   "metadata": {},
   "source": [
    "### Import Python packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f63f31e8-4d2f-4938-adef-601143580f1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import mean_squared_error, r2_score\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from IPython.display import Markdown"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b88e3ca2-011d-4fc6-b645-967958c2609c",
   "metadata": {},
   "source": [
    "### Set the working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0a0f6fc5-bf24-4e8f-bd41-77922667d5da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: /Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run\n"
     ]
    }
   ],
   "source": [
    "# Set the working directory as the parent folder of where the script is located and save it as a variable named \"cwd\"\n",
    "cwd = os.path.dirname(os.getcwd())\n",
    "os.chdir(cwd)\n",
    "\n",
    "# Inspect the current working directory\n",
    "print(f\"Current working directory: {os.getcwd()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "652267d2-883a-49de-941f-6228653eb0b9",
   "metadata": {},
   "source": [
    "### Inspect the bash script to run PRS-CSx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e49ae9ca-d0fe-498a-a89b-696c345040e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bash script path: /Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run/scripts/run_prscsx.sh\n"
     ]
    },
    {
     "data": {
      "text/markdown": [
       "```bash\n",
       "#!/bin/bash\n",
       "\n",
       "## Setting Up PRS-CSx \n",
       "\n",
       "# 0. Set the working directory as the parent directory of which the script is located.\n",
       "echo \"Current working directory: $(realpath ./)\"\n",
       "\n",
       "# 1. Clone the PRS-CSx repository using the following git command:\n",
       "git clone https://github.com/getian107/PRScsx.git\n",
       "\n",
       "# 2. Create a sub-folder named `ref`. Download the LD reference panels to `ref` and extract files:\n",
       "# For regions that don't have access to Dropbox, reference panels can be downloaded from the [alternative download site](https://personal.broadinstitute.org/hhuang/public/PRS-CSx/Reference/).\n",
       "mkdir -p ./inputs/ref\n",
       "echo \"Downloading files...\"\n",
       "EAS_REF_URL=\"https://www.dropbox.com/s/7ek4lwwf2b7f749/ldblk_1kg_eas.tar.gz?dl=1\"\n",
       "EUR_REF_URL=\"https://www.dropbox.com/s/mt6var0z96vb6fv/ldblk_1kg_eur.tar.gz?e=1&dl=0\"\n",
       "EAS_REF_DIR=\"./inputs/ref/ldblk_1kg_eas.tar.gz\"\n",
       "EUR_REF_DIR=\"./inputs/ref/ldblk_1kg_eur.tar.gz\"\n",
       "curl -L -o \"$EAS_REF_DIR\" \"$EAS_REF_URL\"\n",
       "curl -L -o \"$EUR_REF_DIR\" \"$EUR_REF_URL\"\n",
       "echo \"Download completed: 1)EAS ref: $(realpath \"$EAS_REF_DIR\"), 2) EUR ref: $(realpath \"$EUR_REF_DIR\")\"\n",
       "echo \"Extracting files...\"\n",
       "tar -zxvf \"$EAS_REF_DIR\" -C \"./inputs/ref\"\n",
       "tar -zxvf \"$EUR_REF_DIR\" -C \"./inputs/ref\"\n",
       "echo \"Extraction completed.\"\n",
       "\n",
       "# 3. Download the SNP information file and put it in the same folder containing the reference panels:\n",
       "SNP_INFO_URL=\"https://www.dropbox.com/s/rhi806sstvppzzz/snpinfo_mult_1kg_hm3?dl=0\"\n",
       "SNP_INFO_DIR=\"./inputs/ref/snpinfo_mult_1kg_hm3\"\n",
       "curl -L -o \"$SNP_INFO_DIR\" \"$SNP_INFO_URL\"\n",
       "\n",
       "# 4. PRScsx requires Python packages `scipy` and `h5py` installed:\n",
       "# Function to check and install a Python package\n",
       "check_and_install_package() {\n",
       "    PACKAGE=$1\n",
       "    if python -c \"import $PACKAGE\" &> /dev/null; then\n",
       "        echo \"$PACKAGE is already installed.\"\n",
       "    else\n",
       "        echo \"$PACKAGE is not installed. Installing...\"\n",
       "        pip install $PACKAGE\n",
       "        if [ $? -eq 0 ]; then\n",
       "            echo \"$PACKAGE installed successfully.\"\n",
       "        else\n",
       "            echo \"Failed to install $PACKAGE. Please check your Python and pip setup.\"\n",
       "            exit 1\n",
       "        fi\n",
       "    fi\n",
       "}\n",
       "\n",
       "# Ensure pip is available\n",
       "if ! command -v pip &> /dev/null; then\n",
       "    echo \"pip is not installed. Please install pip and rerun this script.\"\n",
       "    exit 1\n",
       "fi\n",
       "\n",
       "# Check and install scipy and h5py\n",
       "check_and_install_package \"scipy\"\n",
       "check_and_install_package \"h5py\"\n",
       "\n",
       "# 5. Once Python and its dependencies have been installed, running the following will print a list of command-line options:\n",
       "echo \"Printing PRScsx command options...\"\n",
       "./PRScsx/PRScsx.py --help \n",
       "\n",
       "## Using PRS-CSx with Test Data\n",
       "# The test data contains EUR and EAS GWAS summary statistics and a bim file for 1,000 SNPs on chromosome 22.\n",
       "# 1. Create a directory to store output:\n",
       "mkdir -p outputs\n",
       "OUTPUT_FOLDER_DIR=$(realpath \"./outputs\")\n",
       "    \n",
       "# 2. Run PRS-CSx:\n",
       "echo \"Running PRS-CSx on test data...\"\n",
       "python ./PRScsx/PRScsx.py \\\n",
       "    --ref_dir=./inputs/ref \\\n",
       "    --bim_prefix=./PRScsx/test_data/test \\\n",
       "    --sst_file=./PRScsx/test_data/EUR_sumstats.txt,./PRScsx/test_data/EAS_sumstats.txt \\\n",
       "    --n_gwas=200000,100000 \\\n",
       "    --pop=EUR,EAS \\\n",
       "    --chrom=22 \\\n",
       "    --phi=1e-2 \\\n",
       "    --out_dir=./outputs \\\n",
       "    --out_name=test\n",
       "echo \"PRS-CSx finished running: $OUTPUT_FOLDER_DIR\"\n",
       "\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Absolute path to the bash script\n",
    "bash_script_path = os.path.join(cwd, \"scripts/run_prscsx.sh\")\n",
    "print(f\"Bash script path: {bash_script_path}\")\n",
    "\n",
    "# Read the script content\n",
    "with open(bash_script_path, \"r\") as file:\n",
    "    bash_script_content = file.read()\n",
    "\n",
    "# Display the script as preformatted Markdown\n",
    "Markdown(f\"```bash\\n{bash_script_content}\\n```\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98262141-af45-4b86-9ea8-304ee5c39b63",
   "metadata": {},
   "source": [
    "### Explanation of the input of PRS-CSx\n",
    "`ref_dir` --> PATH_TO_REFERENCE (required): Full path to the directory that contains the SNP information file and LD reference panels. If the 1000 Genomes reference is used, the folder would contain the SNP information file snpinfo_mult_1kg_hm3 and one or more of the LD reference files: ldblk_1kg_afr, ldblk_1kg_amr, ldblk_1kg_eas, ldblk_1kg_eur, ldblk_1kg_sas; if the UK Biobank reference is used, the folder would contain the SNP information file snpinfo_mult_ukbb_hm3 and one or more of the LD reference files: ldblk_ukbb_afr, ldblk_ukbb_amr, ldblk_ukbb_eas, ldblk_ukbb_eur, ldblk_ukbb_sas. (In our script, we used the 1000 Genomes reference.)\n",
    "\n",
    "`bim_prefix` --> VALIDATION_BIM_PREFIX (required): Full path and the prefix of the bim file for the target (validation/testing) dataset. This file is used to provide a list of SNPs that are available in the target dataset.\n",
    "\n",
    "`sst_file` --> SUM_STATS_FILE (required): Full path and the file name of the GWAS summary statistics. Multiple GWAS summary statistics files are allowed and should be separated by comma. The summary statistics file must include either BETA/OR + SE or BETA/OR + P. The formats are specified on the Github.\n",
    "\n",
    "`n_gwas` --> GWAS_SAMPLE_SIZE (required): Sample sizes of the GWAS, in the same order of the GWAS summary statistics files, separated by comma.\n",
    "\n",
    "`pop` --> POPULATION (required): Population of the GWAS sample, in the same order of the GWAS summary statistics files, separated by comma. For both the 1000 Genomes reference and the UK Biobank reference, AFR, AMR, EAS, EUR and SAS are allowed.\n",
    "\n",
    "`chrom` --> CHROM (optional): The chromosome on which the model is fitted, separated by comma, e.g., --chrom=1,3,5. Parallel computation for the 22 autosomes is recommended. Default is iterating through 22 autosomes (can be time-consuming).\n",
    "\n",
    "`phi` --> PARAM_PHI (optional): Global shrinkage parameter phi. If phi is not specified, it will be learnt from the data using a fully Bayesian approach. This usually works well for polygenic traits with very large GWAS sample sizes (hundreds of thousands of subjects). For GWAS with limited sample sizes (including most of the current disease GWAS), fixing phi to 1e-2 (for highly polygenic traits) or 1e-4 (for less polygenic traits), or doing a small-scale grid search (e.g., phi=1e-6, 1e-4, 1e-2, 1) to find the optimal phi value in the validation dataset often improves perdictive performance.\n",
    "\n",
    "`out_dir` --> OUTPUT_DIR (required): Output directory of the posterior effect size estimates.\n",
    "\n",
    "`out_name` --> OUTPUT_FILE_PREFIX (required): Output filename prefix of the posterior effect size estimates.\n",
    "\n",
    "More optional flags are explained in the Github.\n",
    "\n",
    "Reference: https://github.com/getian107/PRScsx"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d54e5a3a-6a0f-4a3f-a928-2e1022169e39",
   "metadata": {},
   "source": [
    "### Run the bash script to run PRS-CSx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "90c7a9bf-438a-4c98-97e5-e29d5a368ecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run/scripts/run_prscsx.sh\n",
      "Current working directory: /Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run\n",
      "Cloning into 'PRScsx'...\n",
      "remote: Enumerating objects: 276, done.\u001b[K\n",
      "remote: Counting objects: 100% (82/82), done.\u001b[K\n",
      "remote: Compressing objects: 100% (72/72), done.\u001b[K\n",
      "remote: Total 276 (delta 48), reused 22 (delta 10), pack-reused 194 (from 1)\u001b[K\n",
      "Receiving objects: 100% (276/276), 99.90 KiB | 4.76 MiB/s, done.\n",
      "Resolving deltas: 100% (166/166), done.\n",
      "Downloading files...\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   137  100   137    0     0    722      0 --:--:-- --:--:-- --:--:--   724\n",
      "100    17  100    17    0     0     11      0  0:00:01  0:00:01 --:--:--    74\n",
      "100   496    0   496    0     0    224      0 --:--:--  0:00:02 --:--:--     0\n",
      "100 4131M  100 4131M    0     0  15.2M      0  0:04:30  0:04:30 --:--:-- 15.8M:06:10 13.6M  0  11.3M      0  0:06:03  0:00:16  0:05:47 13.2M:00:42  0:04:36 16.2M     0  0:05:07  0:00:58  0:04:09 16.4M2447M    0     0  14.5M      0  0:04:44  0:02:48  0:01:56 15.5M0:01:49 15.3M 14.5M      0  0:04:44  0:03:09  0:01:35 13.9M     0  0:04:42  0:03:15  0:01:27 17.4M0:43 20.9M0  0:04:30  0:03:54  0:00:36 12.2M15.2M      0  0:04:30  0:03:54  0:00:36 11.9M5.1M    0     0  15.2M      0  0:04:31  0:04:19  0:00:12 11.5M31  0:04:29  0:00:02 15.2M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   145  100   140    0     0      0      0 --:--:-- --:--:-- --:--:--     05    0     0    771      0 --:--:-- --:--:-- --:--:--   775\n",
      "100    17  100    17    0     0     20      0 --:--:-- --:--:-- --:--:-- 17000\n",
      "100   491    0   491    0     0    344      0 --:--:--  0:00:01 --:--:--   344\n",
      "100 4352M  100 4352M    0     0  14.8M      0  0:04:53  0:04:53 --:--:-- 15.6M  11.4M      0  0:06:19  0:00:05  0:06:14 17.5M0:04:28  0:00:35  0:03:53 16.4M     0  16.8M      0  0:04:18  0:01:04  0:03:14 17.1M      0  0:04:18  0:01:06  0:03:12 18.1MM      0  0:04:17  0:01:11  0:03:06 17.3M     0  0:04:17  0:01:38  0:02:39 16.4MM    0     0  16.3M      0  0:04:26  0:02:00  0:02:26 12.9M  16.0M      0  0:04:30  0:02:20  0:02:10 14.2M34  0:02:40  0:01:54 14.2MM    0     0  15.6M      0  0:04:37  0:02:49  0:01:48 11.5M5.5M      0  0:04:39  0:02:53  0:01:46 10.8M15.3M      0  0:04:43  0:03:33  0:01:10 14.2M0:04:46  0:04:01  0:00:45 14.7M5.1M      0  0:04:47  0:04:07  0:00:40 14.1M.8M      0  0:04:53  0:04:40  0:00:13 12.2M\n",
      "Download completed: 1)EAS ref: /Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run/inputs/ref/ldblk_1kg_eas.tar.gz, 2) EUR ref: /Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run/inputs/ref/ldblk_1kg_eur.tar.gz\n",
      "Extracting files...\n",
      "x ldblk_1kg_eas/\n",
      "x ldblk_1kg_eas/snpinfo_1kg_hm3\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr1.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr2.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr3.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr4.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr5.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr6.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr7.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr8.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr9.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr10.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr11.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr12.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr13.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr14.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr15.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr16.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr17.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr18.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr19.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr20.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr21.hdf5\n",
      "x ldblk_1kg_eas/ldblk_1kg_chr22.hdf5\n",
      "x ldblk_1kg_eur/\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr1.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr10.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr11.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr12.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr13.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr14.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr15.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr16.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr17.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr18.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr19.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr2.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr20.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr21.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr22.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr3.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr4.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr5.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr6.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr7.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr8.hdf5\n",
      "x ldblk_1kg_eur/ldblk_1kg_chr9.hdf5\n",
      "x ldblk_1kg_eur/snpinfo_1kg_hm3\n",
      "Extraction completed.\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   137  100   137    0     0    891      0 --:--:-- --:--:-- --:--:--   895\n",
      "100    17  100    17    0     0     25      0 --:--:-- --:--:-- --:--:--     0\n",
      "100  100M  100  100M    0     0  13.5M      0  0:00:07  0:00:07 --:--:-- 17.2M\n",
      "scipy is already installed.\n",
      "h5py is already installed.\n",
      "Printing PRScsx command options...\n",
      "\n",
      "\n",
      "\n",
      "PRS-CSx: a method that integrates GWAS summary statistics and external LD reference panels from multiple populations to improve cross-population polygenic prediction.\n",
      "Posterior SNP effect sizes are inferred under coupled continuous shrinkage (CS) priors across populations.\n",
      "PRS-CSx is an extension of the Bayesian polygenic prediction method PRS-CS.\n",
      "\n",
      "References: T Ge, CY Chen, Y Ni, YCA Feng, JW Smoller. Polygenic Prediction via Bayesian Regression and Continuous Shrinkage Priors.\n",
      "           Nature Communications, 10:1776, 2019.\n",
      "\n",
      "           Ruan Y, Lin YF, Feng YCA, Chen CY, Lam M, Guo Z, Stanley Global Asia Initiatives, He L, Sawa A, Martin AR, Qin S, Huang H, Ge T.\n",
      "           Improving Polygenic Prediction in Ancestrally Diverse Populations.\n",
      "           Nature Genetics, 54:573-580, 2022.\n",
      "\n",
      "Usage:\n",
      "python PRScsx.py --ref_dir=PATH_TO_REFERENCE --bim_prefix=VALIDATION_BIM_PREFIX --sst_file=SUM_STATS_FILE --n_gwas=GWAS_SAMPLE_SIZE\n",
      "                 --pop=POPULATION --out_dir=OUTPUT_DIR --out_name=OUTPUT_FILE_PREFIX\n",
      "                 [--a=PARAM_A --b=PARAM_B --phi=PARAM_PHI --n_iter=MCMC_ITERATIONS --n_burnin=MCMC_BURNIN --thin=MCMC_THINNING_FACTOR\n",
      "                  --chrom=CHROM --meta=META_FLAG --write_pst=WRITE_POSTERIOR_SAMPLES --seed=SEED]\n",
      "\n",
      "\n",
      "Running PRS-CSx on test data...\n",
      "\n",
      "\n",
      "--ref_dir=./inputs/ref\n",
      "--bim_prefix=./PRScsx/test_data/test\n",
      "--sst_file=['./PRScsx/test_data/EUR_sumstats.txt', './PRScsx/test_data/EAS_sumstats.txt']\n",
      "--a=1\n",
      "--b=0.5\n",
      "--phi=0.01\n",
      "--n_gwas=[200000, 100000]\n",
      "--pop=['EUR', 'EAS']\n",
      "--n_iter=2000\n",
      "--n_burnin=1000\n",
      "--thin=5\n",
      "--out_dir=./outputs\n",
      "--out_name=test\n",
      "--chrom=['22']\n",
      "--meta=FALSE\n",
      "--write_pst=FALSE\n",
      "--seed=None\n",
      "\n",
      "\n",
      "*** 2 discovery populations detected ***\n",
      "\n",
      "##### process chromosome 22 #####\n",
      "... parse reference file: ./inputs/ref/snpinfo_mult_1kg_hm3 ...\n",
      "... 18944 SNPs on chromosome 22 read from ./inputs/ref/snpinfo_mult_1kg_hm3 ...\n",
      "... parse bim file: ./PRScsx/test_data/test.bim ...\n",
      "... 1000 SNPs on chromosome 22 read from ./PRScsx/test_data/test.bim ...\n",
      "... parse EUR sumstats file: ./PRScsx/test_data/EUR_sumstats.txt ...\n",
      "... 1000 SNPs read from ./PRScsx/test_data/EUR_sumstats.txt ...\n",
      "... 1000 common SNPs in the EUR reference, EUR sumstats, and validation set ...\n",
      "... parse EAS sumstats file: ./PRScsx/test_data/EAS_sumstats.txt ...\n",
      "... 1000 SNPs read from ./PRScsx/test_data/EAS_sumstats.txt ...\n",
      "... 901 common SNPs in the EAS reference, EAS sumstats, and validation set ...\n",
      "... parse EUR reference LD on chromosome 22 ...\n",
      "... parse EAS reference LD on chromosome 22 ...\n",
      "... align reference LD on chromosome 22 across populations ...\n",
      "... 1000 valid SNPs across populations ...\n",
      "... MCMC ...\n",
      "--- iter-100 ---\n",
      "--- iter-200 ---\n",
      "--- iter-300 ---\n",
      "--- iter-400 ---\n",
      "--- iter-500 ---\n",
      "--- iter-600 ---\n",
      "--- iter-700 ---\n",
      "--- iter-800 ---\n",
      "--- iter-900 ---\n",
      "--- iter-1000 ---\n",
      "--- iter-1100 ---\n",
      "--- iter-1200 ---\n",
      "--- iter-1300 ---\n",
      "--- iter-1400 ---\n",
      "--- iter-1500 ---\n",
      "--- iter-1600 ---\n",
      "--- iter-1700 ---\n",
      "--- iter-1800 ---\n",
      "--- iter-1900 ---\n",
      "--- iter-2000 ---\n",
      "... Done ...\n",
      "\n",
      "\n",
      "PRS-CSx finished running: /Users/aliceyan/Documents/GitHub/prs-csx-workshop-tutorial-updated/run_use_evaluate_prscsx/run/outputs\n"
     ]
    }
   ],
   "source": [
    "print(bash_script_path)\n",
    "with open(bash_script_path, \"w\") as file:\n",
    "    file.write(bash_script_content)\n",
    "\n",
    "# Make the script executable\n",
    "!chmod +x ./scripts/run_prscsx.sh\n",
    "\n",
    "# Run the script\n",
    "!./scripts/run_prscsx.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73e137d-9cff-44e3-9ffe-8cbf209bc32e",
   "metadata": {},
   "source": [
    "### Explanation of the output of PRS-CSx\n",
    "*For each input GWAS, PRS-CSx writes `posterior SNP effect size estimates for each chromosome` to the user-specified directory. The output file contains `chromosome, rs ID, base position, A1, A2 and posterior effect size estimate for each SNP`.*\n",
    "\n",
    "*An individual-level polygenic score can be produced by concatenating output files from all chromosomes and then using PLINK's --score command (https://www.cog-genomics.org/plink/1.9/score). If polygenic scores are generated by chromosome, use the 'sum' modifier so that they can be combined into a genome-wide score.* In our tutorial, we are going to demonstrate this calculation process. We will multiply the posterior SNP effect sizes for chromosome 22 by the individual's genotype matrix to compute the PRS.\n",
    "\n",
    "### How to use PRS-CSx?\n",
    "*Given a global shrinkage parameter, the first approach calculates `one polygenic score for each discovery population` using `population-specific posterior SNP effect size estimates` and learns a `linear combination of the polygenic scores` that most accurately predicts the trait in the validation dataset. The optimal global shrinkage parameter and linear combination weights are then taken to an independent dataset, where the predictive performance of the final PRS can be assessed.* The second approach is explained in detail in the Github.\n",
    "\n",
    "Reference: https://github.com/getian107/PRScsx\n",
    "\n",
    "We will domonstrate the usage of PRS-CSx in the following part."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
